### ISSUE TITLE
support M1 macs?

### ISSUE THREAD

**Issue by stebo85** [NeuroDesk/neurodesktop#44]
- we would need to rebuild the desktop image for Arm and ALL containers - feasible?

**Comment by aswinnarayanan** [comment]
**Did look into this a few months back. Tested some arm containers on an m1. Things might have changed now.

The container can technically be rebuilt for arm, but the application would need to support arm for it to work properly
Some of the neuroimaging tools have m1 support, but it looks like rosetta 2 handles them.

This might be a useful resource to track
https://github.com/neurolabusc/AppleSiliconForNeuroimaging

**Comment by stebo85** [comment]
ok, that's great to know. I don't think it anything we can do short-term, but it might be something for a sub-project and we night need dedicated funding to make this happen.

**Comment by stebo85** [comment]
Idea: Use qemu or firecracker to run an x86 Container inside an arm docker container. It would be less performance but a great intermediate solution until the actual application containers are arm compatible.

**Comment by stebo85** [comment]
We may not need to do much except to start an x86 Container on an M1 

https://forums.docker.com/t/run-x86-intel-and-arm-based-images-on-apple-silicon-m1-macs/117123

Can someone test this? I might get my hands on an M1 soon, but let's see!

**Comment by stebo85** [comment]
I might have found a way of running Neurodesk on M1 macs. @aswinnarayanan would be great to hear what you think about it and if it's too crazy:

Docker can run the linux/amd64 image in emulation mode, so most things already work, but one thing in particular is broken: Singularity and apptainer don't work and complain with 

```
singularity shell /cvmfs/neurodesk.ardc.edu.au/containers/fsl_6.0.4_20210105/fsl_6.0.4_20210105.simg/
INFO:    Environment variable SINGULARITY_BINDPATH is set, but APPTAINER_BINDPATH is preferred
INFO:    Environment variable SINGULARITYENV_SUBJECTS_DIR is set, but APPTAINERENV_SUBJECTS_DIR is preferred
ERROR  : Failed to spawn stage 1
```

but: It's possible to run our application containers through chroot and here is a little proof of concept:

```
sudo mount --bind /home/jovyan/ /cvmfs/neurodesk.ardc.edu.au/containers/fsl_6.0.4_20210105/fsl_6.0.4_20210105.simg/root/

sudo chroot /cvmfs/neurodesk.ardc.edu.au/containers/fsl_6.0.4_20210105/fsl_6.0.4_20210105.simg/

export PATH="/opt/fsl-6.0.4/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"

cd 

bet
```
works :p 

Thoughts? Ideas?

**Comment by pdlloyd** [comment]
I'm interested in this issue for a use case I'm working on. It's not Apple Silicon, but I'm trying to get Neurodesktop working on an ARM64 (aarch64) machine with Nvidia GPUs. One option I've considered but haven't tried yet is an x86_64 virtual machine using KVM and performing PCIe pass-through with Virtio. Granted, that'll be a significant performance hit, but maybe Good Enough<sup>TM</sup> until the Singularity containers play a bit nicer. 

Is there any way to get more verbose debugging information from Singularity and Apptainer?

**Comment by pdlloyd** [comment]
Update - I stumbled upon the multiarch container that is presumably built following https://docs.docker.com/build/building/multi-platform/ . I removed `--platform=linux/amd64` and just ran `sudo docker run --rm --shm-size=1gb -it --privileged --user=root --name neurodesktop   -v /home/user/neurodesktop/data:/neurodesktop-storage -e NB_UID="$(id -u)" -e NB_GID="$(id -g)"   -p 8888:8888 ghcr.io/neurodesk/neurodesktop/neurodesktop-multiarch:latest`

No more issues with launching Singularity modules and even when I run `lspci` (having made no additional changes), I see 
```
0a:00.0 3D controller: NVIDIA Corporation GA102GL [A10] (rev a1)
0b:00.0 3D controller: NVIDIA Corporation GA102GL [A10] (rev a1)
```

I don't have a Mac to test on, but it works well on the cloud ARM64 I'm working on.

**Comment by stebo85** [comment]
Dear @pdlloyd,

Yes, we just got the multiarch build to work a few days ago. I am currently testing it on the M1 and with Rosetta enabled in Docker I see an amazing and very usable performance of running x86 singularity containers in an Arm Singularity build. Using Qemu x86 emulation the performance drops by about 80% which I guess is expected. 

How does the x86 translation work on your GPU? Are the application containers actually working and is the emulation also through Qemu?

One thing we want to work on next is to add native Arm containers to this - but we need some funding to do this (or a lot of help!).

keen to hear how you go with your setup!

Thank you
Steffen

**Comment by pdlloyd** [comment]
> One thing we want to work on next is to add native ARM containers to this - but we need some funding to do this (or a lot of help!).

Native ARM containers definitely benefit my application as well. I can't personally put in funding but I would be interested in contributing toward this goal.

> How does the x86 translation work on your CPU? Are the application containers actually working and is the emulation also through Qemu like in MacOS?

The x86 translation on my end happens with [binfmt](https://registry.hub.docker.com/r/tonistiigi/binfmt) which allows QEMU to interpret / emulate the instructions. It didn't seem like an insane performance hit, but I can throw a lot of RAM and cores at it. I increased `shm-size` to `32gb` but I may be misunderstanding how /dev/shm is utilized. Does Neurodesk(top) have any sort of benchmark tools built in? How were you able to determine an 80% performance drop?

I also haven't tried installing drivers or anything for the NVidia cards and using them in-image, which will be interesting to attempt.

**Comment by stebo85** [comment]
Dear @pdlloyd,

Great to hear you would be keen to help with this effort. Which applications would be useful for you to have in native arm? I know that FSL, AFNI and ANTS would be quite straight forward to build on ARM, whereas Freesurfer would be close to impossible.

I wouldn't think that shm-size would bring a lot for x86 emulation on ARM - have you seen any effects of this?

Neurodesk doesn't have a benchmark tool, but I used our example notebooks as a real word benchmark. There I saw the following performance differences between executions, here is an example:
x86 through Rosetta 2 on M1 Pro: 19.49s
x86 through qemu on M1 Pro: 147.31s
x86 on AMD EPYC 7742: 24.51s

As soon as we have a native arm container we could also test how much more performance gain this brings and if it is worth the effort. I also want to repeat these tests on an Ampere processor to get a better feeling for this.

You probably shouldn't try to install the nividia drivers inside the neurodesktop docker container image. I tested a few things there and at least for me it worked better to leave the host deal with nvidia driver and the application containers bring the matching CUDA libraries for the respective applications (like FSL).
